**2 WAY ANOVA:**

there will be two null hypothesis.

--> **Two-Way ANOVA: Assessing the effects of two categorical variables on a continuous outcome**.

In [ ]:

import os

In [ ]:
os.getcwd()

'/content'

In [ ]:
os.chdir(r'/content/drive/MyDrive/STATSCDAC')

In [ ]:
os.getcwd()

'/content/drive/MyDrive/STATSCDAC'

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_excel('CDAC_DataBook.xlsx',sheet_name='salaries')

In [ ]:
import statsmodels
from statsmodels import stats
import statsmodels.api as sm

In [ ]:
from statsmodels.formula.api import ols

In [ ]:
import statsmodels.stats.multicomp

In [ ]:
mydf = df[['rank','gender','salary']]

In [ ]:
mydf.head()

#rank and gender are the predictors.salary is the response or output.

rank gender  salary
0      Prof   Male  139750
1      Prof   Male  173200
2  AsstProf   Male   79750
3      Prof   Male  115000
4      Prof   Male  141500

In [ ]:
mod1 = ols('salary~rank+gender', data=mydf).fit()

In [ ]:
tb1 = sm.stats.anova_lm(mod1)



In [ ]:
print(tb1)

             df        sum_sq       mean_sq           F        PR(>F)
rank        2.0  1.432318e+11  7.161588e+10  128.382480  1.234599e-43
gender      1.0  8.408166e+08  8.408166e+08    1.507293  2.202874e-01
Residual  393.0  2.192281e+11  5.578322e+08         NaN           NaN


establish null hypothesis:

--> H0 for rank:
H0: The average salary is same for all the pairs of ranks.

--> H0 for gender:
H0: The average salary is same for all the pairs of genders.

**In general:**
H0: The average **response** is same for all the pairs of **predictor**.

**For rank:** since the p-value(1.234599e-43) is lessthan 0.05, so we reject the H0. Salary depends on rank.

**For gender:** since the p-value(2.202874e-01) is morethan 0.05, so we DO NOT reject the H0. Salary does not depends on gender.

we're considered both rank and salary, dependency on rank is more and comparing to rank there is negligible dependecy on the gender.

In [ ]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd

In [ ]:
pairwise_tukeyhsd

<function statsmodels.stats.multicomp.pairwise_tukeyhsd(endog, groups, alpha=0.05)>

In [ ]:
mod1 = ols('salary~rank+gender', data=mydf).fit()

In [ ]:
print(pairwise_tukeyhsd(mydf.salary,mydf.gender))

#IMP POINTS!!!!
#Here we're only considered only gender, now  p< 0.05, the gender dependency has significant value. hence now the gender H0 is rejected.
#in the previous case, we're considered both rank and gender. dependecy on rank is much higher than dependecny on gender. hence the
#dependency of gender is negligible compared to dependency of rank.

#during covid wave, there is corruption, other disease cases, but we're more focused on covid and all others are of negligible significance
# compared to covid. but when in normal days the other cases has statistical significance.

    Multiple Comparison of Means - Tukey HSD, FWER=0.05    
group1 group2  meandiff  p-adj    lower     upper    reject
-----------------------------------------------------------
Female   Male 14088.0087 0.0057 4131.1074 24044.9101   True
-----------------------------------------------------------


In [ ]:
tb1 = sm.stats.anova_lm(mod1)
tb1

df        sum_sq       mean_sq           F        PR(>F)
rank        2.0  1.432318e+11  7.161588e+10  128.382480  1.234599e-43
gender      1.0  8.408166e+08  8.408166e+08    1.507293  2.202874e-01
Residual  393.0  2.192281e+11  5.578322e+08         NaN           NaN

INTERACTION BETWEEN TWO FACTORS:

--> here we put * symbol instead of +.
--> between 3 or more factors are not significant generally.


#ex:
Let's say speed is response
this speed depends on 2 factors
1. Mode of tranport - i)bike or ii)car
2. Weather - i)dry or ii)rainy

**Main effect plot:**
two main factors
for each of these two we can make two main effect plots.
response(speed) is on y axis
mode of trnsport on x axis.

case1:
driving bike in rain with speed s1
bike in dry with s2
driving car in rain with speed s3
driving car in dry with speed s4

avg speedof bike without considering weather: s1+s2 /2
avg speedof car without considering weather:  s3+s4 /2

see the graph in your notes.

the line joining the two points of speeds are sm



for this we can draw the interaction plot:

**INTERACTION PLOT:**
take any one factor on any of the axis. and the other one on the another axis.
 WE'RE taking mode of transport on x axis.
 take two different colours for plotting graphs for Rain and Dry(refer notes.)

join the speed points of car and bike w.r.t weather i.e., rain and dry.


IF the two lines are not parallel to each other then the factors have no interaction with each other.

If the change in weather have same impact on bike and car, then the factors have no interaction with each other.

but in real, the effect of rain has more effect on bike than car.
the effect of traffic is more on car than bike.


weather and mode has an interaction because the weather has effect on speed of mode of trnsport.
traffic and mode of transport will have an interaction because the traffic has effect on speed of mode of trnsport.


In [ ]:
mod1 = ols('salary~rank*gender', data=mydf).fit()

In [ ]:
tb1 = sm.stats.anova_lm(mod1)
tb1

df        sum_sq       mean_sq           F        PR(>F)
rank           2.0  1.432318e+11  7.161588e+10  127.754543  2.010089e-43
gender         1.0  8.408166e+08  8.408166e+08    1.499921  2.214209e-01
rank:gender    2.0  4.360306e+07  2.180153e+07    0.038891  9.618588e-01
Residual     391.0  2.191845e+11  5.605741e+08         NaN           NaN

OPTIMIZATION:

In optimization, we will try to get the best maximum output, considering the constraint

In [ ]:
my_std = 45

#sample std dev
#we would like to find lower estimate and upoper estikmate of population std_dev.
#we don't have a single value to the pop std dev because we considered data from sample.

In [ ]:
n = 100

In [ ]:
import scipy

In [ ]:
from scipy import stats

In [ ]:
from scipy.stats import chi2

In [ ]:
crit1 =scipy.stats.chi2.ppf(0.025, 99)

#1st critical value on chi square distribution on X axis.
#0.025 is the area
#99 is degree of freedom

In [ ]:
crit2 = scipy.stats.chi2.ppf(0.975, 99)

In [ ]:
low_est = ((n-1)/crit2)**0.5 * my_std
low_est

39.51030804317418

In [ ]:
upper_est = ((n-1)/crit1)**0.5*my_std
upper_est

52.27538648825794

The pop std dev will fall into this range.

**1 Sample Variace Test:**

--> **This test is used When we wish to compare the std dev of a sample against a claim value to establish whether the difference is statiscially significant or not**.

from above example,
sample std dev =45
pop std dev = 39 to 52

--> If we take sample std dev of 30, it does not falls in the range.  The difference is statisticlly significant.so the H0 is rejected.


--> If we take sample std dev of 40 falls in the range. The difference is statistically not significant. It is subjected to chance variation. the H0 is not rejected.

**TWO SAMPLE VARIANCE TEST:**

This is used to compare the ratio of variances of two samples and whether they are equal or not.

Let's say we have two samples,



In [ ]:
s1= [5,9,8,3,4,7,6,8,9]

In [ ]:
s2= [2,9,8,3,5,7,6,10,12,4]

In [ ]:
np.var(s1, ddof=1)

4.777777777777778

In [ ]:
np.var(s2, ddof=1)

10.266666666666667

In [ ]:
np.var(s1, ddof=1)/np.var(s2, ddof=1)

0.4653679653679653

Null Hypothesis:

H0: both variances are equal OR the ratio of the variances = 1

In [ ]:
from scipy.stats import f

In [ ]:
p_value = (scipy.stats.f.cdf(np.var(s1, ddof=1)/np.var(s2, ddof=1),len(s1)-1, len(s2)-1))*2

#p_value = (scipy.stats.f.cdf(np.var(s1, ddof=1)/np.var(s2, ddof=1),degree of free s1, deg of free s2))*2

In [ ]:
p_value

0.29495626539849623

Statisticlly, no action required since p > 0.05
but here 10.2 and 4.7 is more but according to statistics, it is  not significant. statistics is just guiding you not to take an action. you may take action, it's your wish.

**TOOLS FOR DISCRETE DATA:**

**1) 1 Sample Proportion Test**

--> **In this test, The proportion calculated from a sample has to be compared against a claim value and establish whether the difference is statistically significant or not.**

Eg:

The consultant claims that out of all the people who get jobs on DS, atleast 70% of them are engineers.

H0: prop(engg) >= 0.7
HA: prop(engg) < 0.7

( Null hypothesis can chnage according to the point of view.
the same case above will be reverse for non-engg college.
H0 becomes HA, and HA becomes H0 )

Out of a sample of 270 professionals, 170 are engineers. Does the data support the claim?

The ratio is in decimals, the ratio does not mean that it is continuous data. we consider the sample numbers here for discrete data not the ratio. remember, this point confuses.

In [ ]:
170/270

#sample proportion.

0.6296296296296297

find what are the numbers that fall in range of population proportions.

**--> Std deviation in case of sample proportions is given by**

**std dev** = (p * (1-p) ) **0.5

In [ ]:
s  = (0.6296*(1 - 0.6296))**0.5

#std deviation

In [ ]:
s

#std deviation

0.48291183460337767

--> In proportions we use critical values we get from normal distribution as approximate values.

In [ ]:
low_est = 0.6296 - 1.96*0.4829/270**0.5
low_est

#std dev= 0.4829
#n=270

0.5719988180980312

In [ ]:
upper_est = 0.6296 - 1.96*0.4829/270**0.5
upper_est

0.5719988180980312

--> the population prop for engg is in between low_est= 0.5719988180980312, upper_est = 0.5719988180980312

In [ ]:
import statsmodels
from statsmodels import stats


In [ ]:
from statsmodels.stats import proportion

In [ ]:
from statsmodels.stats.proportion import proportions_ztest

In [ ]:
statsmodels.stats.proportion.proportions_ztest(170,270,0.7,alternative = 'smaller')

(-2.3944789436878944, 0.008321999540908103)

--> p value = 0.008321999540908103

--> p-value is less than 0.05, so we reject the H0. Sample proportion of 0.63 cannot be considered as >= 0.7
the difference is statistically significant.

**2 PROPORTION TEST**

In this test, the difference of proportions of two samples is calculated and compared against the claim value and establish whether the difference is statistically significant or not.

--> My claim is that Banglore has atleast 12% more enginners than Kharghar.

H0= prop(BLR) - prop(KHR) >= 0.12

HA = prop(BLR) - prop(KHR) < 0.12

In BLR, in a sample of 150 students, 125 were engineers

In KHR, in a sample of 100 students, 75 were engineers.



In [ ]:
125/150 - 75/100

0.08333333333333337

--> We need to check whether the difference between 0.083 and 0.12 is statistically significant or not.

--> Here we create two arrays

In [ ]:
ss = np.array([150,100])

#sample space

In [ ]:
en = np.array([125,75])

#no of engineers

In [ ]:
statsmodels.stats.proportion.proportions_ztest(en,ss,0.12,alternative = 'smaller')

#claim value= 0.12

(-0.7100469468046924, 0.23883751199295206)

--> p value = 0.24 > 0.05,

--> so the H0 is not rejected

--> the difference between 0.12 and 0.083 is not significant

--> the difference is due to chance variation

--> no action is required.